packages/common/src/interfaces.ts

```diff
@@ -15,26 +15,43 @@ export type WorkflowQueryType = (...args: any[]) => any;
  */
 export type Workflow = (...args: any[]) => WorkflowReturnType;
 
+declare const argsBrand: unique symbol;
 /**
  * An interface representing a Workflow signal definition, as returned from {@link defineSignal}
  *
- * @remarks `_Args` can be used for parameter type inference in handler functions and *WorkflowHandle methods.
+ * @remarks `Args` can be used for parameter type inference in handler functions and *WorkflowHandle methods.  
+ * `Name` can optionally be specified with a string literal type to preserve type-level knowledge of the signal name.
  */
-// eslint-disable-next-line @typescript-eslint/no-unused-vars
-export interface SignalDefinition<_Args extends any[] = []> {
+export interface SignalDefinition<Args extends any[] = [], Name extends string = string> {
   type: 'signal';
-  name: string;
+  name: Name;
+  /**
+   * Virtual type brand to maintain a distinction between {@link SignalDefinition} types with different args.
+   * This field is not present at run-time.
+   */
+  [argsBrand]: Args;
 }
 
+declare const retBrand: unique symbol;
```

Comment:
should this be exported?

Reply:
No, it's intentionally not exported - the idea here is that unlike patching `__brand: T` onto the type, `[retBrand]: T` is _inaccessable_: callers cannot attempt to do `myQueryDef[retBrand]`, expect to get `Ret`, and get `undefined`, because they cannot access the [Symbol](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol) needed to look up the key.

This is also why I've used `declare const retBrand: unique symbol` instead of `const retBrand = Symbol()`; the idea is that this key a) exists only in the type system, not at run time and b) you cannot even attempt to read it. It's just a type discriminator.

Reply:
That makes sense.